### ISSUE TITLE
No warning for unreachable code after throw expression

### ISSUE THREAD

**Issue by marek-safar** [dotnet/roslyn#19984]
**Version Used**: 2.3.0.61719 (5fbdd5c5)

**Steps to Reproduce**:

```csharp
using System;

class X
{
    public static void Main ()
    {
        var x = true ? throw new NullReferenceException () : 1;
        x = 2; // This code is unreachable but no info about it
        return;
    }
}
```

**Expected Behavior**:

Proper warning about unreachable code

**Actual Behavior**:

No warning.

**Comment by sharwell** [comment]
:memo: The equivalent code using an `if` statement does produce a warning:

```csharp
int x;
if (true)
{
  throw new NullReferenceException();
}
else
{
  x = 1;
}

x = 2; // warning is reported here
```

**Comment by CyrusNajmabadi** [comment]
There is no language in the C# spec about reachability of subexpressions, only reachability of statements.  So this is not an actual bug/issue with Roslyn.

It is, however, a perfectly reasonable request for the language to support this concept.  As such, this should be moved to https://github.com/dotnet/csharplang and recreated as a discussion there.

**Comment by CyrusNajmabadi** [comment]
Closing this out.  Feel free to reopen as a discussion on csharplang.

**Comment by sharwell** [comment]
It is my understanding that a warning should be reported on the `x = 2;` line according to the current specification, augmented by the available specification for throw expressions.

**From the specification**

> A warning is reported if the compiler determines that a statement is unreachable.

> To determine whether a particular statement or end point is reachable, the compiler performs flow analysis according to the reachability rules defined for each statement. The flow analysis takes into account the values of constant expressions (Constant expressions) that control the behavior of statements...

**From the available specification for throw expressions:**

> the rest of the semantics of the throw expression are identical to the semantics of the *throw_statement* in the current language specification

**From the specification for *throw_statement:**

> Because a `throw` statement unconditionally transfers control elsewhere, the end point of a `throw` statement is never reachable.

**Questionable parts:**

There is no discussion of reachability in the specification for *declaration_statement*. The closest would be *expression_statement*, which contains the following language:

> Execution of an *expression_statement* evaluates the contained expression and then transfers control to the end point of the *expression_statement*. The end point of an *expression_statement* is reachable if that *expression_statement* is reachable.

However, it is not possible for a program to adhere to both the specification for throw expression and the language above. I believe it is reasonable to assume that the language added for throw expressions having reachability as defined by throw_statement implies that conflicting language in older constructs will be updated to adhere to the new semantics.

**Comment by CyrusNajmabadi** [comment]
Yes. I can see how that interpretation would hold.  Tagging @gafter @MadsTorgersen for their interpretations here.

**Comment by CyrusNajmabadi** [comment]
So, reading the spec, i think the current compiler implementation is correct.  Looking at expression statements (since local declarations don't even talk about reachability) they simply say:

> Execution of an expression-statement evaluates the contained expression and then transfers control to the end point of the expression-statement. The end point of an expression-statement is reachable if that expression-statement is reachable

In this case, the statement is reachable, so the endpoint is simply considered reachable.  

--

Note: i'm not saying that the initial request is incorrect.  I think the language should have the concepts of expression-reachability, and that throw-expressions should participate in that.  But it feels like a language proposal for csharplang and not an issue for dotnet/roslyn.

**Comment by sharwell** [comment]
@CyrusNajmabadi You are looking at the specification for C# prior to throw expressions. I was considering the merged result of that and the throw expressions specification, and applied a conflict resolution.

**Comment by CyrusNajmabadi** [comment]
Yeha, but the conflict resolution would not apply (based on my reading).  There simply is no thing that changes the endpoint reachability of expression-statements or local delcaration statements.  If the statement is reachable, their endpoint is reachable.  The only way that would change is if their endpoint reachability was dependent on their nested expression end-point being reachable.  However, as there is still nothing in the spec about expression endpoint reachability, nothing connects the two.

**Comment by CyrusNajmabadi** [comment]
>My post includes a description of how the language you quoted is not compatible with the available text regarding throw expressions.

I think it's still fully compatible.  Basically, you're looking at:

> the rest of the semantics of the throw expression are identical to the semantics of the throw_statement in the current language specification

But there simply is no semantics around expression reachability.  It's literally not a concept the language has.  So you can't take the semantics around statement reachability and just apply it to expressions.  

Even if we wanted to apply throw-reachability to apply to expressions, it still wouldn't apply to ternary expressions.  That's because the language says nothing about reachability of ternary expressions.  

--

Again, this is a good idea, but there is a large jump happening here that would only come from direct changes to the language spec.  We'd need *at the minimum* the following: something stating that the endpoint of an expression-statement/local-declaration being reachable if the end-point of it's expression was reachable.  And something stating the reachability of ternary expressions + constant expressions.  

*If* we had those clauses, then we could take the part you quoted about semantics being identical and then take the throw-statemnet reachability semantics and attempt to apply them to throw-expression reachability.

**Comment by marek-safar** [comment]
@CyrusNajmabadi consider following sample (it's just same code shaped differently)

```csharp
using System;

class X
{
    int TestExpr (out int z)
    {
        return true ? throw new NullReferenceException () : 1;
    }
}
```

Now if C# compiler would really follow your statement of

> There is no language in the C# spec about reachability of subexpressions, only reachability of statements. So this is not an actual bug/issue with Roslyn.

then this code should fail to compile because `z` is not always assigned to before control leaves the current method but it compiles fine.

**Comment by sharwell** [comment]
@CyrusNajmabadi I agree that the spec needs to be *much* clearer in this regard. I was simply hoping to see this proceed based on a desire to justify the implementation of a helpful user feature. It's not unlike how throw expressions themselves were implemented without actually being part of the spec.

If you'd like, I can prepare an initial pull request to update [throw expressions](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.0/throw-expression.md) to describe the intended impact on end-point reachability of declaration and expression statements, which is really just the interpretation of the existing text such that this issue can be implemented.

**Comment by CyrusNajmabadi** [comment]
Please just open a bug on CSharpLang so that we actually spec this out and approve this as an actual language change.  It sounds like you're interested in this work so I think we could argue you could be the champion here, and you coudl def provide a PR as part of that championing.

**Comment by CyrusNajmabadi** [comment]
> then this code should fail to compile 

Yes.  This sounds like a bug if it is not failing to compile.

**Comment by gafter** [comment]
No, it should compile without any problem. The definite assignment specification for a throw expression says that every variable is definitely assigned after a throw expression. Not a bug.

**Comment by CyrusNajmabadi** [comment]
Ah, gotcha.  That is per-spec then.

**Comment by sharwell** [comment]
I filed this bug:
dotnet/csharplang#656

If we can agree on the **Intent** section, then this issue can proceed while the details are yet to be worked out in full.

**Comment by marek-safar** [comment]
@gafter The `z` is not definitely assigned at the end of conditional expression in my second example. It's only that (the unspecified) reachability rules alter the definite assignment to not to consider the expression to be conditional expression.

**Comment by gafter** [comment]
> The `z` is not definitely assigned at the end of conditional expression in my second example.

Yes, it is. That is because it is definitely assigned on both branches of the conditional - on the left branch because every variable is definitely assigned after a throw expression, and on the right branch because every variable is "definitely assigned when false" for a constant expression whose value is true (the condition's expression).

**Comment by jcouv** [comment]
Per @gafter's comment this is by design. I see that a csharplang issue was filed.
I'll go ahead and close OP. Thanks
